datastore/adapter/mount.py

```diff
@@ -0,0 +1,231 @@
+			return False
+		return await ds.contains(subkey)
+	
+	
+	async def stat(self, key: datastore.Key) -> MD:
+		ds, subkey = self._find_mountpoint(key)
+		if ds is None:
+			raise KeyError(key)
+		return await ds.stat(subkey)  # type: ignore[return-value] # noqa: F723
+	
+	
+	def mount(self, prefix: datastore.Key, ds: DS) -> None:
+		"""Mounts the datastore `ds` at key `prefix`
+		
+		If a datastore is already mounted at the given key a :exc:`KeyError` is
+		raised.
+		"""
+		current:  _Adapter.MOUNTS_T = self.mounts
+		previous: _Adapter.MOUNTS_T
+		
+		# Walk and create all parent key parts
+		for part in map(str, prefix.list):
+			if part not in current:
+				current[part] = ({}, None)
+			previous = current
+			current  = current[part][0]
+		
+		# Add the given datastore as final entry value
+		entry = previous[str(prefix.list[-1])]
+		if entry[1] is not None:
+			raise KeyError(prefix)
+		previous[str(prefix.list[-1])] = (entry[0], ds)
+	
+	
+	def unmount(self, prefix: datastore.Key) -> DS:
+		"""Unmounts and returns the datastore at key `prefix`
+		
+		If no datastore is mounted at the given key a :exc:`KeyError` is
+		raised.
+		
+		The returned datastore is not closed; it is the callers responsibility
+		to ensure this by using ``await m.unmount(key).aclose()`` or similar.
+		"""
+		current: _Adapter.MOUNTS_T = self.mounts
+		
+		# Walk and create all parent key parts
+		visited: typing.List[_Adapter.MOUNTS_T] = []
+		for part in map(str, prefix.list):
+			if part not in current:
+				raise KeyError(prefix)
+			visited.append(current)
+			current = current[part][0]
+		
+		# Remove the datastore from the final key part
+		entry = visited[-1][str(prefix.list[-1])]
+		if entry[1] is None:
+			raise KeyError(prefix)
+		visited[-1][str(prefix.list[-1])] = (entry[0], None)
+		
+		# Remove all now-empty parents
+		for current in reversed(visited):
+			if list(current.values()) == [({}, None)]:
+				current.clear()
+		
+		return entry[1]
+	
+	
+	def unmount_all(self) -> typing.List[DS]:
+		"""Unmounts and returns all datastores currently mounted
+		
+		The returned datastores are not closed; it is the callers responsibility
+		to ensure this. For removing __and__ closing all mounts use the
+		:meth:`aclose` method instead.
+		"""
+		unmounted: typing.List[DS] = []
+		
+		# Start at mount hierarchy root
+		current: _Adapter.MOUNTS_T = self.mounts
+		
+		# Walk and create all parent key parts
+		stack: typing.List[_Adapter.MOUNTS_T] = []
+		while len(self.mounts) > 0 or len(stack) > 0:
+			try:
+				# Pop item from submounts map
+				_, (next, ds) = current.popitem()
+				
+				# Move datastore (if any) to list of unmounted datastores
+				if ds is not None:
+					unmounted.append(ds)
+				
+				# Recurse into submount
+				stack.append(current)
+				current = next
+			except KeyError:
+				# Move out of empty submount (this assignment drops the last
+				# reference to the submount)
+				current = stack.pop()
+		
+		assert self.mounts == {}
```

Comment:
Is it better to have an `AssertionError` here over an explicit `raise Exception`?